set.mm

```diff
@@ -28582,23 +28567,165 @@ atomic formula (class version of ~ elsb1 ).  Usage of this theorem is
+       (Proof modification is discouraged.)  (New usage is discouraged.) $)
+    ralrexbidOLD $p |- ( A. x e. A ph -> ( E. x e. A ps <-> E. x e. A th ) ) $=
+      ( wral cv wcel wa wex wrex wi wal wb df-ral imim2i pm5.32d alexbii df-rex
+      sylbi 3bitr4g ) ADEGZDHEIZBJZDKZUDCJZDKZBDELCDELUCUDAMZDNUFUHOADEPUIUEUGD
+      UIUDBCABCOUDFQRSUABDETCDETUB $.
+  $}
+
+  $( Extend wff notation to include restricted existential uniqueness. $)
+  wreu $a wff E! x e. A ph $.
+
+  $( Extend wff notation to include restricted "at most one". $)
+  wrmo $a wff E* x e. A ph $.
+
+  $( Extend class notation to include the restricted class abstraction (class
+     builder). $)
+  crab $a class { x e. A | ph } $.
+
+  $( Define restricted "at most one".
+
+     Note:  This notation is most often used to express that ` ph ` holds for
+     at most one element of a given class ` A ` .  For this reading ` F/_ x A `
+     is required, though, for example, asserted when ` x ` and ` A ` are
+     disjoint.
+
+     Should instead ` A ` depend on ` x ` , you rather assert at most one ` x `
+     fulfilling ` ph ` happens to be contained in the corresponding
+     ` A ( x ) ` .  This interpretation is rarely needed (see also ~ df-ral ).
+     (Contributed by NM, 16-Jun-2017.) $)
+  df-rmo $a |- ( E* x e. A ph <-> E* x ( x e. A /\ ph ) ) $.
+
+  $( Define restricted existential uniqueness.
+
+     Note:  This notation is most often used to express that ` ph ` holds for
+     exactly one element of a given class ` A ` .  For this reading ` F/_ x A `
+     is required, though, for example, asserted when ` x ` and ` A ` are
+     disjoint.
+
+     Should instead ` A ` depend on ` x ` , you rather assert exactly one ` x `
+     fulfilling ` ph ` happens to be contained in the corresponding
+     ` A ( x ) ` .  This interpretation is rarely needed (see also ~ df-ral ).
+     (Contributed by NM, 22-Nov-1994.) $)
+  df-reu $a |- ( E! x e. A ph <-> E! x ( x e. A /\ ph ) ) $.
+
+  $( Define a restricted class abstraction (class builder): ` { x e. A | ph } `
+     is the class of all sets ` x ` in ` A ` such that ` ph ( x ) ` is true.
+     Definition of [TakeutiZaring] p. 20.
+
+     For the interpretation given in the previous paragraph to be correct, we
+     need to assume ` F/_ x A ` , which is the case as soon as ` x ` and ` A `
+     are disjoint, which is generally the case.  If ` A ` were to depend on
+     ` x ` , then the interpretation would be less obvious (think of the two
+     extreme cases ` A = { x } ` and ` A = x ` , for instance).  See also
+     ~ df-ral .  (Contributed by NM, 22-Nov-1994.) $)
+  df-rab $a |- { x e. A | ph } = { x | ( x e. A /\ ph ) } $.
+
+  ${
+    2ralimi.1 $e |- ( ph -> ps ) $.
+    $( Inference quantifying both antecedent and consequent two times, with
+       strong hypothesis.  (Contributed by AV, 3-Dec-2021.) $)
+    2ralimi $p |- ( A. x e. A A. y e. B ph -> A. x e. A A. y e. B ps ) $=
+      ( wral ralimi ) ADFHBDFHCEABDFGII $.
+  $}
```

Comment:
not sure why 2ralimi et al is right after df-rmo df-reu df-rab, instead of somewhere before. In my opinion I would put it after ralimi/reximi, I know this conflicts with the one dimensional before multi dimensional rule, but I feel like connecting related theorems where one dimensional and multi dimensional are in a series would take precedence.

Reply:
[alral](https://us.metamath.org/mpeuni/alral.html) depends on ax-4.